### PR TITLE
Drag and drop ( Media library )

### LINKED ISSUE #54: Discovery: Drag and Drop / Paste Media upload
## Feature description

Broken off from the original ticket ampproject/amp-wp#3805 . This ticket is the functionality that allows users to drag and drop rich media, such as images, videos and audio files into the media library. 

## Acceptance criteria

1. After dragging media from the desktop to the media library it should be added to the library.
1. After dragging media from the desktop to the stage/page it should be added to the library and stage/page.

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### LINKED ISSUE #228: Drag and drop (Library / Media Panel)
As a user I want to be able to upload media files to the library via drag and drop

[Feature Brief - Epic: Media Gallery](https://docs.google.com/document/d/1aZH_TI_be2t5Ywyqx4jcPTPdhFGEt7xh-uLzBp2AV4A/edit#)

**Acceptance Criteria**

1. A user is able to drag media from their desktop into the media library. 
2. A user is able to drag multiple media elements (of mixed type) from their desktop into the media library.
3. Any media that is dragged into the Media Panel that is instantly uploaded should show up instantly in the panel.
4. If the media is not uploaded immediately, the [uploading UI](https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=2683%3A0) is shown. #231 
5. If the upload fails, an error message is shown
6. A drop target is shown on the Panel when hovering with files to drop

**Design**
https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=2683%3A0

### PR DESCRIPTION
**Pull request by spacedmonkey**
Fix #228

### CODE REVIEW

**Comment by wassgha** [comment]
Let's use draft PRs for WIPs, I think it's cleaner

**Comment by spacedmonkey** [comment]
@wassgha I am unable to change it now I have created it. Can you change it for me?

**Comment by swissspidy** [comment]
I don't think GitHub supports changing back to draft. It's fine like this for me 👍

**Comment by spacedmonkey** [comment]
This PR is blocked by #195

**Comment by spacedmonkey** [comment]
This ticket is now blocked #210

File: assets/src/edit-story/components/draganddrop/index.js
Comment (swissspidy) [inline assets/src/edit-story/components/draganddrop/index.js]: Perhaps at some point we could use something like `wp_sprintf_l()` in JS for strings like this.

Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: @swissspidy By default, webp and svg will not be supported here. Should it be remove from this message?

File: assets/src/edit-story/components/draganddrop/icons/upload.svg
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/icons/upload.svg]: Should the `fill` be replaced with `currentColor`?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: Capitalize "And" here and elsewhere if we decide to keep "drag" part here.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: That's just a `default` export? E.g. `import UploadIcon from './...svg';`?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: Where is the `56px` coming from?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: Maybe "Overlay" since it's conventionally a single word?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: To clarify: this code does not deal with `components/dropzone` at all, does it? Regardless of whether we want to reuse this or not, this component looks like more of a "DropZone". We do not appear to dragging anything out of it at this time, only "into".

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: Do we need `onDragStart` here? IMHO, if any drag start is present - it should be handled by a specific component. This seems more of a drop target.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: Do we need to cancel defaults in `onDragLeave`?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: For `onDragEnter`, `onDragOver` and `onDrop` - we need to check the media type, etc, to decide whether we want to allow the drop, right?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: `const files = [...dt.files]`?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: We should only print what we can support at this point. Everything else we should file in an issue and assign to @pbakaus :)

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: It's missing `position: relative` if we using `position:absolute` inside.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/draganddrop/index.js]: Let's use the new `StoryPropTypes.children`.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/components/draganddrop/icons/upload.svg
Comment (barklund) [reply assets/src/edit-story/components/draganddrop/icons/upload.svg]: Yes, set `fill="currentColor"`. Also remove the `width` and `height` from the `<svg />` element (but preserve `viewBox`) to allow for scaling.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (barklund) [reply assets/src/edit-story/components/draganddrop/index.js]: Unfortunately not. Because we've switched to svgr (which is better), the default import is now the file path as a string - and you have to do named imports to get the component. It's a bit annoying, but [can't be helped](https://github.com/google/web-stories-wp/pull/244#issuecomment-583114369).

Comment (swissspidy) [reply assets/src/edit-story/components/draganddrop/index.js]: Well, #244 hasn't been merged yet though.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [inline assets/src/edit-story/components/draganddrop/index.js]: If this is specific to the media library let's have it inside `components/library` instead

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [inline assets/src/edit-story/components/draganddrop/index.js]: Let's rename this to `DropZone` and remove (or temporarily rename) the component in `./components/dropzone` once reorderable is implemented for layers and thumbnails

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [reply assets/src/edit-story/components/draganddrop/index.js]: `const files = [...evt.dataTransfer.files]` since `dt` is not used anywhere else

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [inline assets/src/edit-story/components/draganddrop/index.js]: rename this to `<DropzoneOverlay>` and `< DragandDropOverLay>` is not needed (see comment above)

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [inline assets/src/edit-story/components/draganddrop/index.js]: Let's use `${({theme}) => theme.colors.bg.v1}`

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [reply assets/src/edit-story/components/draganddrop/index.js]: If this is to offset for the tab bar, sync with Sam on whether it's actually better to cover the tab bar as well. It's not going to be clickable anyways as you're dragging and it looks a bit odd uncovered. Otherwise, let's have the dropzone wrap the content of the tab instead of the entire panel and get rid of the `56px`

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [inline assets/src/edit-story/components/draganddrop/index.js]: Let's use flex box on `DragandDropOverLayWrapper` to center the content and get rid of this extra layer
```
display: flex;
flex-direction: column;
justify-content: center;
align-items: center;
```

File: assets/src/edit-story/components/draganddrop/index.js
Comment (wassgha) [inline assets/src/edit-story/components/draganddrop/index.js]: I'm assuming this means that the overlay only appears when dragging into the media library, I think we need to make it appear regardless of where the user is dragging so that we can guide them visually to drop it on the media library.

**Review by wassgha — DISMISSED** [review]

File: assets/src/edit-story/components/draganddrop/index.js
Comment (swissspidy) [reply assets/src/edit-story/components/draganddrop/index.js]: We should make this string dynamic depending on the list of allowed mime types from the ConfigProvider. Might need to reword it a little bit to make that easier. For example:

```
The following file types are supported:
%s
```

which then will be turned into:

```
The following file types are supported:
png, jpg, jpeg, gif, xyz, abc
```

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: In the design I used to implement, the tab bar was visual. I personally done like this, so can change it back to `0` as it has changed in the design anyway.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: This check happens [here](https://github.com/google/web-stories-wp/blob/f8689299965987595e229d9cb2bbd764858da01e/assets/src/edit-story/app/uploader/useUploader.js#L66). We are awaiting #232 to do this properly.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: What is the issue with doing this?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Drag starting did not work in my testing, so I disabled it.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Dropping into the canvas is different and I think should be a different on component IMO. Most of the heavily lifting in this component is done in the uploader hook anymore

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: I did as I saw the change in #244  . I hoped it would land before this PR went in. But if it doesn't then I am willing to change it back to the way I had it before.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: IMHO we should cover it all. But if they want to keep it showing, I believe we need to place the DND overlay in the right container.

Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: I.e., yes, it should be `0`.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: That are basically two different sides of the same coin. The `onDragStart` is for the source (an element with `draggable="true"` attribute). The `onDragEnter`, `onDragOver`, `onDrop`, etc - are all for the drop target - the dragged element is intended to be dropped. For us the drag source is always outside of our editor: desktop or another app/site. So we are really only deal with the "drop" side of things.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: Probably shouldn't be anything problematic, but not very meaningful in this context, imho. We cancel defaults to indicate that the drop is allowed on this element. But there's no drop once you leave...

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: Hmm. Typically you'd want to ignore an unsupported data altogether. E.g. it's usually:

```
if (isValidType()) {
  event.preventDefault();
}
```

This is done so that the other handlers could potentially handle this higher up the DOM. But I wonder if your way is actually is better for our use case since the default browser action is to open whatever is dropped on it. It's so bad that maybe matching the drop target and throwing an error is actually better. Ok. Let's do it this the way you have for now.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Now that #244 is merged, making as resolved.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/1b18686d1ee9e5f698955249a0247d14f13334bd

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Renamed and moved https://github.com/google/web-stories-wp/pull/175/commits/1b18686d1ee9e5f698955249a0247d14f13334bd

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Moved in https://github.com/google/web-stories-wp/pull/175/commits/1b18686d1ee9e5f698955249a0247d14f13334bd

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: This didnt work in my testing. I used position absolute to break the container.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: That did not work, as it is a solid color, this as transparency.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Changed in https://github.com/google/web-stories-wp/pull/175/commits/1b18686d1ee9e5f698955249a0247d14f13334bd

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/cce35b1409450e194144e31f377a75f5247886d3

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: I think you drag a file or files in here, we should show an error message and not let the browser perform it's default behavior.

File: assets/src/edit-story/components/draganddrop/icons/upload.svg
Comment (swissspidy) [reply assets/src/edit-story/components/draganddrop/icons/upload.svg]: Thanks to #244 this _should_ now happen automatically during build.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/e38006343fad70f6ea68cdfa4828282ddc65fdad

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Good catch - https://github.com/google/web-stories-wp/pull/175/commits/b24143f75c42b4c94c8ea32a85c2ea6fd1b23962

File: assets/src/edit-story/components/draganddrop/icons/upload.svg
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/icons/upload.svg]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/f9242e2890a45c6af9ad7424a4b8726022dd99ca

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Changed in https://github.com/google/web-stories-wp/pull/175/commits/1b18686d1ee9e5f698955249a0247d14f13334bd

File: assets/src/edit-story/components/draganddrop/index.js
Comment (swissspidy) [reply assets/src/edit-story/components/draganddrop/index.js]: You can `import { rgba } from 'polished';`

and then `background-color: ${rgba(ttheme.colors.bg.v1, 0.6)};`

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/b3c946741838699c90eeff6baa214814dab10537

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: @samitron7 Can you confirm ^

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: @swissspidy @dvoytenko What do you think of this - https://github.com/google/web-stories-wp/pull/175/commits/b4327cdd620561e12ef080b1a6bbf4d77da2a1e7

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Good catch @swissspidy . Fixed in https://github.com/google/web-stories-wp/pull/175/commits/537f224f77bf5b46114d20caa3849c27f6043236

File: assets/src/edit-story/components/draganddrop/index.js
Comment (swissspidy) [reply assets/src/edit-story/components/draganddrop/index.js]: Thanks! 👍 

I'd just optimize the PHP part a little bit:

```php
/**
 * Returns a flat list of allowed file types (e.g. jpeg, gif, png, mp4).
 *
 * @return array List of allowed file types.
 */
protected static function get_allowed_file_types() {
	$allowed_mime_types = self::get_allowed_mime_types();
	$mime_types         = [];

	foreach ( $allowed_mime_types as $key => $types ) {
		array_push( $mime_types, ...$types );
	}

	$allowed_file_types = [];
	$all_mime_types       = wp_get_mime_types();

	foreach ( $all_mime_types as $key => $type ) {
		if ( in_array( $type, $mime_types, true ) ) {
			array_push( $allowed_file_types, ...explode( '|', $key ) );
		}
	}

	return $allowed_file_types;
}
```

* Make variable names more clear, fix typos
* Avoid `array_merge` in loops

Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/53eb38484c2578bb14635e25862cb5f1aac17944

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: 👍

File: assets/src/edit-story/components/draganddrop/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/draganddrop/index.js]: It looks like external dropping into canvas is mostly the same at this time. I.e. dnd from desktop. The only difference is that after upload we also place the element in the canvas. But totally ok to leave it for a follow up PRs.

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/library/dropzone/index.js:83]: I think we need to set the allowed operation to "copy" here.

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/library/dropzone/index.js]: I still think this needs to be removed.

**Review by dvoytenko — APPROVED** [review]
LGTM on my side with two small requests. Otherwise since there are a lot of comments, please get at least one more LGTM before merging.

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/dropzone/index.js:83]: Can you explain this? Not sure I know what you mean.

Comment (swissspidy) [reply assets/src/edit-story/components/library/dropzone/index.js:83]: I think he is referring to [`DataTransfer.effectAllowed`](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/effectAllowed):

```js
evt.dataTransfer.effectAllowed = 'copy';
```

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/dropzone/index.js]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/29f4b783f1c739af55f8f40561f0826e34f67bb0

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/dropzone/index.js:83]: Good catch. Changed in https://github.com/google/web-stories-wp/pull/175/commits/9360b1c286c77270c665a940fa33fda3ba9573c3

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/dropzone/index.js:124]: Needs a translator comment

File: includes/Story_Post_Type.php
Comment (swissspidy) [inline includes/Story_Post_Type.php:323]: Reminds me that we should add some much needed tests for this class. Opened #328 for this.

**Comment by swissspidy** [comment]
@spacedmonkey Is it just me or does inserting media elements to the page not work anymore now?

File: assets/src/edit-story/components/draganddrop/index.js
Comment (swissspidy) [reply assets/src/edit-story/components/draganddrop/index.js]: Didn't we decide that it should only show when dragging into the media library? Otherwise it should show the canvas dropzone.

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/library/dropzone/index.js:83]: yeah. this one. i seem to remember that in some cases/browsers the `ondrop` might be not called unless this is set.

File: assets/src/edit-story/components/draganddrop/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/draganddrop/index.js]: This is being confirmed by @jauyong

File: assets/src/edit-story/components/library/dropzone/index.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/dropzone/index.js:124]: Fixed in https://github.com/google/web-stories-wp/pull/175/commits/af654de8d1b45833657e9712e306937bd8c89459

**Comment by spacedmonkey** [comment]
> @spacedmonkey Is it just me or does inserting media elements to the page not work anymore now?

🤦‍♂ That is why I don't push last thing on a Friday. I tried something clever with mouse events and css and it worked a little too well it seems. Fix can be found here - https://github.com/google/web-stories-wp/pull/175/commits/af654de8d1b45833657e9712e306937bd8c89459

File: includes/Story_Post_Type.php
Comment (swissspidy) [inline includes/Story_Post_Type.php]: Unrelated change, and also wrong as per https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters

File: includes/Story_Post_Type.php
Comment (swissspidy) [inline includes/Story_Post_Type.php]: Unrelated change here.

(NB: doesn't PHPCS cover this? 🤔)

**Review by swissspidy — APPROVED** [review]
Two minor nits but otherwise 👍

File: includes/Story_Post_Type.php
Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: PHPCS linter went weird. Fixed https://github.com/google/web-stories-wp/pull/175/commits/ba7d5fc86d484dee22161941caf619a6432f414b

File: includes/Story_Post_Type.php
Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: PHPCS went weird. https://github.com/google/web-stories-wp/pull/175/commits/ba7d5fc86d484dee22161941caf619a6432f414b

**Pull request merged**
